File: scripts/module_resolver.py
#!/usr/bin/env python3
"""
module_resolver.py - the canonical way to resolve a module descriptor.

Phase 3 Step 2 of docs/platform-closure-plan.md. Full contract in
docs/reports/module-descriptor-contract.md. This resolver reads
modules/<key>/module.json (or, when a caller explicitly overrides the
search root - fixture callers only - <root>/<key>/module.json) and
validates it against the canonical schema.

Nothing that currently reads modules/yoma paths directly is migrated onto
this yet (that is Steps 3A-4). This module is self-contained: it does not
import worker_pipeline.py, and nothing in worker_pipeline.py imports it
yet either.

No implicit default module. Every caller passes an explicit key. No
caller in this file ever falls back to "yoma".
"""
import json
import re
from pathlib import Path
from types import MappingProxyType

REPO = Path(__file__).resolve().parent.parent
DEFAULT_SEARCH_ROOT = REPO / "modules"

KEY_RE = re.compile(r"^[a-z][a-z0-9_-]*$")

REQUIRED_TOP_FIELDS = (
    "key", "displayNameEn", "status", "publishable", "dafRange", "totalDaf",
    "paths", "schemaMapRef", "capabilities", "buildRuntime",
)
REQUIRED_PATH_FIELDS = (
    "root", "scriptsRoot", "sourceAssetsRoot", "generatedAssetsRoot",
    "sourceStore", "learningDataDir", "learningDataFile", "coverageFile",
)
VALID_STATUS = ("production", "synthetic")


class ModuleResolutionError(Exception):
    """Raised for every resolution failure. .code is one of:
    INVALID_KEY, UNKNOWN_MODULE, MALFORMED_DESCRIPTOR, MISSING_FIELD,
    FEATURE_INCONSISTENCY.
    """

    def __init__(self, code, message):
        self.code = code
        super().__init__(f"{code}: {message}")


class ModuleDescriptor:
    """Read-only wrapper around a validated descriptor dict."""

    def __init__(self, data):
        self._data = MappingProxyType(dict(data))

    def __getitem__(self, item):
        return self._data[item]

    def get(self, item, default=None):
        return self._data.get(item, default)

    def to_dict(self):
        return dict(self._data)

    @property
    def key(self):
        return self._data["key"]

    @property
    def status(self):
        return self._data["status"]

    @property
    def publishable(self):
        return self._data["publishable"]

    @property
    def is_synthetic(self):
        return self._data["status"] == "synthetic"

    @property
    def paths(self):
        return self._data["paths"]

    @property
    def capabilities(self):
        return self._data["capabilities"]

    def __repr__(self):
        return f"ModuleDescriptor(key={self.key!r}, status={self.status!r})"


def _require(cond, code, message):
    if not cond:
        raise ModuleResolutionError(code, message)


def _validate_capability(caps, name, required_config_fields):
    entry = caps.get(name)
    if entry is None:
        raise ModuleResolutionError(
            "MISSING_FIELD", f"capabilities.{name}")
    _require(isinstance(entry, dict), "MALFORMED_DESCRIPTOR",
              f"capabilities.{name} must be an object")
    _require("enabled" in entry and isinstance(entry["enabled"], bool),
              "MISSING_FIELD", f"capabilities.{name}.enabled")
    present_config = [f for f in required_config_fields if entry.get(f) is not None]
    if entry["enabled"]:
        missing = [f for f in required_config_fields if f not in present_config]
        _require(not missing, "FEATURE_INCONSISTENCY",
                  f"capabilities.{name}.enabled is true but missing: "
                  f"{', '.join('capabilities.' + name + '.' + f for f in missing)}")
    else:
        _require(not present_config, "FEATURE_INCONSISTENCY",
                  f"capabilities.{name}.enabled is false but declares "
                  f"{', '.join('capabilities.' + name + '.' + f for f in present_config)} "
                  "anyway - a disabled feature must not carry its own config")


SOURCE_ACQUISITION_STRATEGIES = ("remote-fetch", "local-fixture")


def _validate_source_acquisition(caps):
    """Unlike rashi/literalTranslation, every module has SOME source-
    acquisition strategy - there is no legal "disabled" state, so this is
    not modeled as an enabled/disabled capability. 'remote-fetch' (Yoma's
    strategy: live talmud.dev/Sefaria calls, run interactively/offline by
    an operator, never by CI) requires sourceSystem and fetchScript.
    'local-fixture' (synthetic modules only, e.g. the Phase 3 fixture)
    requires fixtureInputDir and forbids any network access - enforced
    here only as a schema requirement; the actual no-network guarantee is
    a property of what script the module points fixtureInputDir/fetchScript
    at, which this resolver cannot verify by reading JSON."""
    sa = caps.get("sourceAcquisition")
    _require(isinstance(sa, dict), "MISSING_FIELD", "capabilities.sourceAcquisition")
    strategy = sa.get("strategy")
    _require(strategy in SOURCE_ACQUISITION_STRATEGIES, "MALFORMED_DESCRIPTOR",
              f"capabilities.sourceAcquisition.strategy must be one of "
              f"{SOURCE_ACQUISITION_STRATEGIES}, got {strategy!r}")
    if strategy == "remote-fetch":
        _require(sa.get("sourceSystem") and sa.get("fetchScript"),
                  "MISSING_FIELD",
                  "capabilities.sourceAcquisition.sourceSystem and .fetchScript "
                  "(required when strategy is remote-fetch)")
        _require(not sa.get("fixtureInputDir"), "FEATURE_INCONSISTENCY",
                  "capabilities.sourceAcquisition.fixtureInputDir is set but "
                  "strategy is remote-fetch, not local-fixture")
    else:  # local-fixture
        _require(sa.get("fixtureInputDir"), "MISSING_FIELD",
                  "capabilities.sourceAcquisition.fixtureInputDir "
                  "(required when strategy is local-fixture)")
        _require(not (sa.get("sourceSystem") or sa.get("fetchScript")),
                  "FEATURE_INCONSISTENCY",
                  "capabilities.sourceAcquisition.sourceSystem/.fetchScript are "
                  "set but strategy is local-fixture, not remote-fetch - a "
                  "synthetic module must not claim a live source system")


def validate_descriptor(data, key):
    """Validate a parsed descriptor dict for the given expected key.

    Raises ModuleResolutionError on any violation. Returns nothing on
    success - callers wrap the dict in ModuleDescriptor themselves.
    """
    _require(isinstance(data, dict), "MALFORMED_DESCRIPTOR",
              "descriptor root must be a JSON object")

    for field in REQUIRED_TOP_FIELDS:
        _require(field in data and data[field] is not None,
                  "MISSING_FIELD", field)

    _require(data["key"] == key, "MALFORMED_DESCRIPTOR",
              f"descriptor key {data['key']!r} does not match its own "
              f"directory key {key!r}")
    _require(bool(KEY_RE.match(data["key"])), "MALFORMED_DESCRIPTOR",
              f"key {data['key']!r} is not a valid machine-safe slug "
              f"(must match {KEY_RE.pattern})")

    _require(data["status"] in VALID_STATUS, "MALFORMED_DESCRIPTOR",
              f"status must be one of {VALID_STATUS}, got {data['status']!r}")
    _require(isinstance(data["publishable"], bool), "MALFORMED_DESCRIPTOR",
              "publishable must be a boolean")
    if data["status"] == "synthetic":
        _require(data["publishable"] is False, "FEATURE_INCONSISTENCY",
                  "status is synthetic but publishable is true - a "
                  "synthetic fixture may never be publishable")

    _require(isinstance(data["totalDaf"], int) and data["totalDaf"] > 0,
              "MALFORMED_DESCRIPTOR", "totalDaf must be a positive integer")

    if data["dafRange"] is not None:
        dr = data["dafRange"]
        _require(isinstance(dr, dict) and "first" in dr and "last" in dr,
                  "MALFORMED_DESCRIPTOR",
                  "dafRange must be null or {first, last}")

    paths = data["paths"]
    _require(isinstance(paths, dict), "MALFORMED_DESCRIPTOR",
              "paths must be an object")
    for field in REQUIRED_PATH_FIELDS:
        _require(field in paths and paths[field], "MISSING_FIELD", f"paths.{field}")

    expected_root = f"modules/{key}"
    _require(paths["root"] == expected_root, "MALFORMED_DESCRIPTOR",
              f"paths.root {paths['root']!r} does not match the expected "
              f"module root {expected_root!r} for key {key!r}")
    for field in ("scriptsRoot", "sourceAssetsRoot", "generatedAssetsRoot",
                  "sourceStore", "learningDataDir", "learningDataFile",
                  "coverageFile"):
        p = paths[field]
        _require(isinstance(p, str) and not p.startswith("/") and ".." not in p.split("/"),
                  "MALFORMED_DESCRIPTOR",
                  f"paths.{field} must be a relative, non-traversing path, got {p!r}")
        _require(p == paths["root"] or p.startswith(paths["root"] + "/"),
                  "FEATURE_INCONSISTENCY",
                  f"paths.{field} ({p!r}) is not under the module's own "
                  f"root ({paths['root']!r})")

    caps = data["capabilities"]
    _require(isinstance(caps, dict), "MALFORMED_DESCRIPTOR",
              "capabilities must be an object")
    _validate_capability(caps, "rashi", ("allowlistsRoot",))
    _validate_capability(caps, "literalTranslation", ("assetsDir",))
    _validate_source_acquisition(caps)

    br = data["buildRuntime"]
    _require(isinstance(br, dict) and br.get("dataScript"),
              "MISSING_FIELD", "buildRuntime.dataScript")
    _require(br["dataScript"] == paths["learningDataFile"],
              "FEATURE_INCONSISTENCY",
              f"buildRuntime.dataScript ({br['dataScript']!r}) disagrees "
              f"with paths.learningDataFile ({paths['learningDataFile']!r}) "
              "- both name the same file and must not drift")


def resolve_module(key, search_root=None):
    """Resolve and validate one module descriptor.

    key: the explicit module identifier. Required; there is no default.
    search_root: directory to look for <key>/module.json under. Defaults
        to modules/ (production discovery). Fixture/test callers pass an
        explicit override (e.g. tests/fixtures/modules/); production code
        paths never do.

    Returns a ModuleDescriptor. Raises ModuleResolutionError on any
    failure - never silently substitutes another module.
    """
    if not isinstance(key, str) or not KEY_RE.match(key):
        raise ModuleResolutionError(
            "INVALID_KEY",
            f"{key!r} is not a valid module key (must match {KEY_RE.pattern})")

    root = Path(search_root) if search_root is not None else DEFAULT_SEARCH_ROOT
    descriptor_path = root / key / "module.json"

    # Defense in depth: even though KEY_RE already forbids '/', '..', and
    # backslashes, explicitly confirm the resolved path stays under root
    # before any read, so a future regex loosening can't silently reopen
    # traversal.
    try:
        resolved = descriptor_path.resolve()
        root_resolved = root.resolve()
    except OSError as e:  # pragma: no cover - defensive
        raise ModuleResolutionError("MALFORMED_DESCRIPTOR", str(e))
    if root_resolved not in resolved.parents and resolved != root_resolved:
        raise ModuleResolutionError(
            "INVALID_KEY", f"resolved path {resolved} escapes search root {root_resolved}")

    if not descriptor_path.is_file():
        raise ModuleResolutionError(
            "UNKNOWN_MODULE", f"no descriptor at {descriptor_path}")

    try:
        raw = descriptor_path.read_text(encoding="utf-8")
    except OSError as e:
        raise ModuleResolutionError("MALFORMED_DESCRIPTOR", str(e))

    try:
        data = json.loads(raw)
    except json.JSONDecodeError as e:
        raise ModuleResolutionError("MALFORMED_DESCRIPTOR", f"invalid JSON: {e}")

    validate_descriptor(data, key)
    return ModuleDescriptor(data)


def list_modules(search_root=None):
    """Return the sorted list of module keys discoverable under search_root
    (production default: modules/). A directory whose module.json fails
    validation is skipped, not raised - callers that need strict behavior
    should call resolve_module() on each key themselves."""
    root = Path(search_root) if search_root is not None else DEFAULT_SEARCH_ROOT
    if not root.is_dir():
        return []
    keys = []
    for child in sorted(root.iterdir()):
        if child.is_dir() and (child / "module.json").is_file():
            keys.append(child.name)
    return keys


if __name__ == "__main__":
    import argparse
    import sys

    ap = argparse.ArgumentParser(description=__doc__.split("\n")[1])
    ap.add_argument("key", help="module key to resolve, e.g. yoma")
    ap.add_argument("--search-root", default=None)
    args = ap.parse_args()

    try:
        d = resolve_module(args.key, search_root=args.search_root)
    except ModuleResolutionError as e:
        print(f"ERROR  {e.code}: {e}", file=sys.stderr)
        sys.exit(1)
    print(json.dumps(d.to_dict(), indent=2, ensure_ascii=False))

File: shared/module_resolver.js
/**
 * module_resolver.js - the canonical way to resolve a module descriptor,
 * JS side (for build.mjs and future .mjs tools, migrated onto this
 * starting Phase 3 Step 4).
 *
 * Full contract: docs/reports/module-descriptor-contract.md. Independent
 * re-implementation of scripts/module_resolver.py's validation rules -
 * deliberately not shared code across languages, so a bug in one is not
 * automatically a bug in the other.
 *
 * No implicit default module. Every caller passes an explicit key.
 *
 * Dual-usable like the rest of shared/: a plain script with no import
 * side effects when read directly, plus a CommonJS export guard for
 * Node tooling.
 */
(function (root, factory) {
  if (typeof module !== "undefined" && module.exports) {
    module.exports = factory(require("fs"), require("path"));
  } else {
    root.ModuleResolver = factory(null, null);
  }
})(typeof self !== "undefined" ? self : this, function (fs, path) {
  "use strict";

  var KEY_RE = /^[a-z][a-z0-9_-]*$/;
  var VALID_STATUS = ["production", "synthetic"];
  var REQUIRED_TOP_FIELDS = [
    "key", "displayNameEn", "status", "publishable", "dafRange", "totalDaf",
    "paths", "schemaMapRef", "capabilities", "buildRuntime",
  ];
  var REQUIRED_PATH_FIELDS = [
    "root", "scriptsRoot", "sourceAssetsRoot", "generatedAssetsRoot",
    "sourceStore", "learningDataDir", "learningDataFile", "coverageFile",
  ];

  function ModuleResolutionError(code, message) {
    var err = new Error(code + ": " + message);
    err.code = code;
    err.name = "ModuleResolutionError";
    return err;
  }

  function requireCond(cond, code, message) {
    if (!cond) throw ModuleResolutionError(code, message);
  }

  function validateCapability(caps, name, requiredConfigFields) {
    var entry = caps[name];
    requireCond(entry !== undefined && entry !== null, "MISSING_FIELD",
      "capabilities." + name);
    requireCond(typeof entry === "object" && !Array.isArray(entry),
      "MALFORMED_DESCRIPTOR", "capabilities." + name + " must be an object");
    requireCond(typeof entry.enabled === "boolean", "MISSING_FIELD",
      "capabilities." + name + ".enabled");
    var present = requiredConfigFields.filter(function (f) {
      return entry[f] !== undefined && entry[f] !== null;
    });
    if (entry.enabled) {
      var missing = requiredConfigFields.filter(function (f) {
        return present.indexOf(f) === -1;
      });
      requireCond(missing.length === 0, "FEATURE_INCONSISTENCY",
        "capabilities." + name + ".enabled is true but missing: " +
        missing.map(function (f) { return "capabilities." + name + "." + f; }).join(", "));
    } else {
      requireCond(present.length === 0, "FEATURE_INCONSISTENCY",
        "capabilities." + name + ".enabled is false but declares " +
        present.map(function (f) { return "capabilities." + name + "." + f; }).join(", ") +
        " anyway - a disabled feature must not carry its own config");
    }
  }

  var SOURCE_ACQUISITION_STRATEGIES = ["remote-fetch", "local-fixture"];

  function validateSourceAcquisition(caps) {
    // Unlike rashi/literalTranslation, every module has SOME
    // source-acquisition strategy - there is no legal "disabled" state,
    // so this is not modeled as an enabled/disabled capability.
    var sa = caps.sourceAcquisition;
    requireCond(typeof sa === "object" && sa !== null, "MISSING_FIELD",
      "capabilities.sourceAcquisition");
    var strategy = sa.strategy;
    requireCond(SOURCE_ACQUISITION_STRATEGIES.indexOf(strategy) !== -1,
      "MALFORMED_DESCRIPTOR",
      "capabilities.sourceAcquisition.strategy must be one of " +
      SOURCE_ACQUISITION_STRATEGIES.join(", ") + ", got " + strategy);
    if (strategy === "remote-fetch") {
      requireCond(!!sa.sourceSystem && !!sa.fetchScript, "MISSING_FIELD",
        "capabilities.sourceAcquisition.sourceSystem and .fetchScript " +
        "(required when strategy is remote-fetch)");
      requireCond(!sa.fixtureInputDir, "FEATURE_INCONSISTENCY",
        "capabilities.sourceAcquisition.fixtureInputDir is set but " +
        "strategy is remote-fetch, not local-fixture");
    } else {
      requireCond(!!sa.fixtureInputDir, "MISSING_FIELD",
        "capabilities.sourceAcquisition.fixtureInputDir " +
        "(required when strategy is local-fixture)");
      requireCond(!(sa.sourceSystem || sa.fetchScript), "FEATURE_INCONSISTENCY",
        "capabilities.sourceAcquisition.sourceSystem/.fetchScript are set " +
        "but strategy is local-fixture, not remote-fetch - a synthetic " +
        "module must not claim a live source system");
    }
  }

  function validateDescriptor(data, key) {
    requireCond(typeof data === "object" && data !== null && !Array.isArray(data),
      "MALFORMED_DESCRIPTOR", "descriptor root must be a JSON object");

    REQUIRED_TOP_FIELDS.forEach(function (field) {
      requireCond(data[field] !== undefined && data[field] !== null,
        "MISSING_FIELD", field);
    });

    requireCond(data.key === key, "MALFORMED_DESCRIPTOR",
      "descriptor key " + JSON.stringify(data.key) + " does not match its " +
      "own directory key " + JSON.stringify(key));
    requireCond(KEY_RE.test(data.key), "MALFORMED_DESCRIPTOR",
      "key " + JSON.stringify(data.key) + " is not a valid machine-safe slug");

    requireCond(VALID_STATUS.indexOf(data.status) !== -1, "MALFORMED_DESCRIPTOR",
      "status must be one of " + VALID_STATUS.join(", ") + ", got " + data.status);
    requireCond(typeof data.publishable === "boolean", "MALFORMED_DESCRIPTOR",
      "publishable must be a boolean");
    if (data.status === "synthetic") {
      requireCond(data.publishable === false, "FEATURE_INCONSISTENCY",
        "status is synthetic but publishable is true - a synthetic " +
        "fixture may never be publishable");
    }

    requireCond(Number.isInteger(data.totalDaf) && data.totalDaf > 0,
      "MALFORMED_DESCRIPTOR", "totalDaf must be a positive integer");

    if (data.dafRange !== null) {
      requireCond(typeof data.dafRange === "object" &&
        "first" in data.dafRange && "last" in data.dafRange,
        "MALFORMED_DESCRIPTOR", "dafRange must be null or {first, last}");
    }

    var paths = data.paths;
    requireCond(typeof paths === "object" && paths !== null, "MALFORMED_DESCRIPTOR",
      "paths must be an object");
    REQUIRED_PATH_FIELDS.forEach(function (field) {
      requireCond(!!paths[field], "MISSING_FIELD", "paths." + field);
    });

    var expectedRoot = "modules/" + key;
    requireCond(paths.root === expectedRoot, "MALFORMED_DESCRIPTOR",
      "paths.root " + JSON.stringify(paths.root) + " does not match the " +
      "expected module root " + JSON.stringify(expectedRoot));

    ["scriptsRoot", "sourceAssetsRoot", "generatedAssetsRoot", "sourceStore",
      "learningDataDir", "learningDataFile", "coverageFile"].forEach(function (field) {
      var p = paths[field];
      requireCond(typeof p === "string" && p.charAt(0) !== "/" &&
        p.split("/").indexOf("..") === -1, "MALFORMED_DESCRIPTOR",
        "paths." + field + " must be a relative, non-traversing path, got " + p);
      requireCond(p === paths.root || p.indexOf(paths.root + "/") === 0,
        "FEATURE_INCONSISTENCY",
        "paths." + field + " (" + p + ") is not under the module's own root (" + paths.root + ")");
    });

    var caps = data.capabilities;
    requireCond(typeof caps === "object" && caps !== null, "MALFORMED_DESCRIPTOR",
      "capabilities must be an object");
    validateCapability(caps, "rashi", ["allowlistsRoot"]);
    validateCapability(caps, "literalTranslation", ["assetsDir"]);
    validateSourceAcquisition(caps);

    var br = data.buildRuntime;
    requireCond(typeof br === "object" && br !== null && !!br.dataScript,
      "MISSING_FIELD", "buildRuntime.dataScript");
    requireCond(br.dataScript === paths.learningDataFile, "FEATURE_INCONSISTENCY",
      "buildRuntime.dataScript (" + br.dataScript + ") disagrees with " +
      "paths.learningDataFile (" + paths.learningDataFile + ") - both name " +
      "the same file and must not drift");
  }

  /**
   * resolveModule(key, repoRoot, searchRoot?) -> descriptor object (frozen)
   *
   * key: explicit module identifier, required, no default.
   * repoRoot: absolute path to the repository root (Node-only; required
   *   when fs/path are available).
   * searchRoot: absolute path to search under; defaults to
   *   <repoRoot>/modules (production discovery). Fixture/test callers
   *   pass an explicit override; production code paths never do.
   *
   * Throws ModuleResolutionError (Error with .code) on any failure.
   * Never silently substitutes another module.
   */
  function resolveModule(key, repoRoot, searchRoot) {
    if (typeof key !== "string" || !KEY_RE.test(key)) {
      throw ModuleResolutionError("INVALID_KEY",
        JSON.stringify(key) + " is not a valid module key (must match " + KEY_RE + ")");
    }
    if (!fs || !path) {
      throw ModuleResolutionError("MALFORMED_DESCRIPTOR",
        "resolveModule requires a Node fs/path environment");
    }
    var root = searchRoot || path.join(repoRoot, "modules");
    var descriptorPath = path.join(root, key, "module.json");

    var resolvedPath = path.resolve(descriptorPath);
    var resolvedRoot = path.resolve(root);
    var rel = path.relative(resolvedRoot, resolvedPath);
    if (rel.startsWith("..") || path.isAbsolute(rel)) {
      throw ModuleResolutionError("INVALID_KEY",
        "resolved path " + resolvedPath + " escapes search root " + resolvedRoot);
    }

    if (!fs.existsSync(descriptorPath) || !fs.statSync(descriptorPath).isFile()) {
      throw ModuleResolutionError("UNKNOWN_MODULE",
        "no descriptor at " + descriptorPath);
    }

    var raw;
    try {
      raw = fs.readFileSync(descriptorPath, "utf-8");
    } catch (e) {
      throw ModuleResolutionError("MALFORMED_DESCRIPTOR", String(e));
    }

    var data;
    try {
      data = JSON.parse(raw);
    } catch (e) {
      throw ModuleResolutionError("MALFORMED_DESCRIPTOR", "invalid JSON: " + e.message);
    }

    validateDescriptor(data, key);
    return Object.freeze(JSON.parse(JSON.stringify(data)));
  }

  function listModules(repoRoot, searchRoot) {
    var root = searchRoot || path.join(repoRoot, "modules");
    if (!fs || !fs.existsSync(root)) return [];
    return fs.readdirSync(root)
      .filter(function (name) {
        var full = path.join(root, name);
        return fs.statSync(full).isDirectory() &&
          fs.existsSync(path.join(full, "module.json"));
      })
      .sort();
  }

  return {
    resolveModule: resolveModule,
    listModules: listModules,
    validateDescriptor: validateDescriptor,
    ModuleResolutionError: ModuleResolutionError,
    KEY_RE: KEY_RE,
  };
});
